File: app/src/main/java/br/fmu/bioreino/ui/activity/MainActivity.java
package br.fmu.bioreino.ui.activity;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.widget.ImageButton;
import android.widget.Toast;
import android.widget.Toolbar;

import androidx.activity.OnBackPressedCallback;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import com.google.android.material.navigation.NavigationView;

import br.fmu.bioreino.ui.fragments.AulasFragment;
import br.fmu.bioreino.R;
import br.fmu.bioreino.model.Curso;
import br.fmu.bioreino.ui.fragments.CategoriaFragment;
import br.fmu.bioreino.ui.fragments.HomeFragment;
import br.fmu.bioreino.util.Comunicador;

public class MainActivity extends AppCompatActivity implements Comunicador {

    private final FragmentManager fragmentManager = getSupportFragmentManager();
    private DrawerLayout drawerLayout;

    // Fragments -----------------------------------------------------------------------------------
    HomeFragment homeFragment = new HomeFragment();
    AulasFragment aulasFragment = new AulasFragment();
    CategoriaFragment categoriaFragment = new CategoriaFragment();

    // Métodos principais --------------------------------------------------------------------------
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        configuraMenu();

        trocaTela(homeFragment);
    }

    // Configurações -------------------------------------------------------------------------------
    private void configuraMenu() {
        drawerLayout = findViewById(R.id.activity_main_drawerLayout);
        configuraBotaoAppbar();
        configuraMenuNavegation();
    }

    private void configuraMenuNavegation() {
        NavigationView menu = findViewById(R.id.activity_home_navigationView);
        menu.inflateMenu(R.menu.menu_navigation);
        menu.setNavigationItemSelectedListener(item -> {

            final int home = R.id.menu_navigation_home;

            int id = item.getItemId();
            drawerLayout.closeDrawer(GravityCompat.START);
            item.setChecked(true);

            switch (id) {

                case home:
                    trocaTela(homeFragment);
                    break;

                default:
                    mostraMensagemAindaNaoDisponivel(this);
            }

            return false;
        });
    }



    // Helpers -------------------------------------------------------------------------------------
    @Override
    public void trocaTela(Fragment fragment) {
        FragmentTransaction fragmentTransaction = getFragmentTransaction(fragment);

        fragmentTransaction.replace(R.id.activity_home_frameLayout, fragment);
        fragmentTransaction.commit();
    }

    @Override
    public void trocaTela(Fragment fragment, Bundle bundle) {
        FragmentTransaction fragmentTransaction = getFragmentTransaction(fragment);

        fragment.setArguments(bundle);

        fragmentTransaction.replace(R.id.activity_home_frameLayout, fragment);
        fragmentTransaction.commit();
    }

    @NonNull
    private FragmentTransaction getFragmentTransaction(Fragment fragment) {
        return (fragment != homeFragment)
                ? fragmentManager.beginTransaction().addToBackStack(fragment.getTag())
                : fragmentManager.beginTransaction();
    }

    @Override
    public void enviaCurso(Curso curso) {
        Bundle bundle = new Bundle();
        bundle.putSerializable("curso", curso);

        trocaTela(aulasFragment, bundle);
    }

    @Override
    public void enviaCategoria(String categoria) {
        Bundle bundle = new Bundle();
        bundle.putString("categoria", categoria);

        trocaTela(categoriaFragment, bundle);
    }

    @Override
    public void trocaAppBar(Toolbar toolbar) {

    }

    @Override
    public void configuraBotaoVoltarSistema(Fragment fragmentAtual) {
        OnBackPressedCallback quandoBotaoVoltarPressionado = new OnBackPressedCallback(true) {
            @Override
            public void handleOnBackPressed() {
                fragmentAtual.getParentFragmentManager().popBackStack();
            }
        };
        fragmentAtual
                .requireActivity()
                .getOnBackPressedDispatcher()
                .addCallback(fragmentAtual, quandoBotaoVoltarPressionado);
    }


    public void mostraMensagemAindaNaoDisponivel(Context context) {
        Toast.makeText(context, "Ainda não disponível", Toast.LENGTH_SHORT).show();
    }

    private void configuraBotaoAppbar() {
        ImageButton botaoMenu = findViewById(R.id.activity_home_appbar_menu);
        botaoMenu.setOnClickListener(view -> drawerLayout.openDrawer(GravityCompat.START));
    }

}

File: app/src/main/java/br/fmu/bioreino/view/activity/MainActivity.kt
package br.fmu.bioreino.view.activity

import android.content.Context
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import android.widget.Toolbar
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.forEach
import androidx.drawerlayout.widget.DrawerLayout
import androidx.drawerlayout.widget.DrawerLayout.DrawerListener
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentTransaction
import br.fmu.bioreino.LoadingFragment
import br.fmu.bioreino.R
import br.fmu.bioreino.controller.util.Comunicator
import br.fmu.bioreino.databinding.ActivityMainBinding
import br.fmu.bioreino.model.entities.Curso
import br.fmu.bioreino.view.fragments.CategoriesFragment
import br.fmu.bioreino.view.fragments.HomeFragment
import br.fmu.bioreino.view.fragments.LessonsFragment
import com.google.android.material.navigation.NavigationView

class MainActivity : AppCompatActivity(), Comunicator {

    private val mainBinding by lazy {
        ActivityMainBinding.inflate(layoutInflater)
    }

    private val fragmentManager = supportFragmentManager
    private val homeFragment = HomeFragment()
    private val lessonsFragment = LessonsFragment()
    private val categoriesFragment = CategoriesFragment()

    private lateinit var drawer: DrawerLayout

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(mainBinding.root)
        configurateMenu()
        changeFragment(homeFragment)

    }


    // Configurations ------------------------------------------------------------------------------
    private fun configurateMenu() {
        drawer = findViewById(R.id.activity_main_drawerLayout)
        openDrawerWith(R.id.activity_home_appbar_menu)
        configurateDrawerNavigation()
    }

    // Configs -------------------------------------------------------------------------------------
    private fun openDrawerWith(viewId: Int) {
        val view = findViewById<View>(viewId)
        view.setOnClickListener { drawer.open() }
    }

    private fun configurateDrawerNavigation() {
        val drawerMenu: NavigationView = mainBinding.activityMainNavigationView
        val homeId = R.id.menu_navigation_home
        drawerMenu.inflateMenu(R.menu.navigation_menu)
        drawerMenu.menu.findItem(homeId).isChecked = true
        drawerMenu.setNavigationItemSelectedListener {
            val itemId = it.itemId
            drawerMenu.menu.forEach { item: MenuItem ->
                if (item.hasSubMenu()) item.subMenu?.forEach { subItem ->
                    subItem.isChecked = false
                }
            }

            it.isChecked = true
            when (itemId) {
                homeId -> changeFragmentFromMenu(homeFragment)
                else -> showNotYetDefined(this)
            }


            true
        }

    }


    // Comunicator ---------------------------------------------------------------------------------
    override fun changeFragment(fragment: Fragment, bundle: Bundle?) {
        val fragmentTransaction: FragmentTransaction = getFragmentTransaction(fragment)
        bundle.let { fragment.arguments = bundle }
        fragmentTransaction.replace(R.id.activity_home_frameLayout, fragment, "actual_fragment")
        fragmentTransaction.commit()
    }

    override fun enviaCurso(curso: Curso) {
        val bundle = Bundle()
        bundle.putSerializable("curso", curso)
        changeFragment(lessonsFragment, bundle)
    }

    override fun enviaCategoria(categoria: String) {
        val bundle = Bundle()
        bundle.putString("categoria", categoria)
        changeFragment(categoriesFragment, bundle)
    }

    override fun trocaAppBar(toolbar: Toolbar) {
        TODO("Not yet implemented")
    }

    override fun configuraBotaoVoltarSistema(fragmentAtual: Fragment) {
        val quandoBotaoVoltarPressionado: OnBackPressedCallback =
            object : OnBackPressedCallback(true) {
                override fun handleOnBackPressed() {
                    fragmentAtual.parentFragmentManager.popBackStack()
                }
            }
        fragmentAtual.requireActivity().onBackPressedDispatcher.addCallback(
            fragmentAtual,
            quandoBotaoVoltarPressionado
        )
    }


    // Helpers -------------------------------------------------------------------------------------
    private fun getFragmentTransaction(fragment: Fragment): FragmentTransaction {
        return if (fragment !== homeFragment) {
            fragmentManager.beginTransaction().addToBackStack(fragment.tag)
        } else {
            fragmentManager.beginTransaction()
        }
    }

    private fun changeFragmentFromMenu(fragment: Fragment) {
        fragmentManager.findFragmentByTag("actual_fragment")?.let {
            if (it != fragment) {
                changeFragment(LoadingFragment())
                drawer.addDrawerListener(object : DrawerListener {
                    override fun onDrawerClosed(drawerView: View) {
                        changeFragment(fragment)
                    }
                    override fun onDrawerSlide(drawerView: View, slideOffset: Float) {}
                    override fun onDrawerOpened(drawerView: View) {}
                    override fun onDrawerStateChanged(newState: Int) {}
                })
            }
            drawer.close()
        }

    }

    fun showNotYetDefined(context: Context?) {
        Toast.makeText(context, "Ainda não disponível", Toast.LENGTH_SHORT).show()
    }

}